longformer/longformer.py

```diff
@@ -204,8 +200,15 @@ def forward(
             # (bsz, seq_len, num_heads, extra attention count + 2*window+1)
             attn_weights = torch.cat((selected_attn_weights, attn_weights), dim=-1)
         if getattr(self, 'global_tokens', 0) > 0:
+            q_g = self.query_global(hidden_states)
+            k_g = self.key_global(hidden_states)
+            v_g = self.value_global(hidden_states)
+            q_g = q_g.contiguous().view(seq_len, bsz, self.num_heads, self.head_dim).transpose(0, 1)
+            k_g = k_g.contiguous().view(seq_len, bsz, self.num_heads, self.head_dim).transpose(0, 1)
+            v_g = v_g.contiguous().view(seq_len, bsz, self.num_heads, self.head_dim).transpose(0, 1)
+
             # (bsz, seq_len, num_heads, max_num_extra_indices_per_batch)
-            selected_attn_weights = torch.einsum('blhd,bshd->blhs', (q, k[:, :self.global_tokens]))
+            selected_attn_weights = torch.einsum('blhd,bshd->blhs', (q_g, k_g[:, :self.global_tokens]))
```

Comment:
This block of code should be moved below. When every token is attending to the global tokens, we use regular projection (this 
![image](https://user-images.githubusercontent.com/2287797/98181316-c6e9ed80-1eb7-11eb-9602-16d33babff64.png)).

The global projections are used when computing attention from global tokens to all tokens (this 
![Uploading image.png…]())

Reply:
This clarifies it 👌